File: qtr_pairing_process/board_analysis.py
"""Board-level outlook: what is guaranteed, what is reachable, what is at stake.

Motivation
----------
The engine has always scored a pairing decision with a single number derived
from how good the board *can* be. Measured against real event data that number
ties across candidate decisions on 100% of boards, because it reports the same
ceiling for many different choices (``docs/WTC2024_GROUND_TRUTH.md``, Finding
13). What separates those choices is the **floor** -- how bad the round can
still become once the decision is made.

Two facts from WTC 2024 motivate the design:

* **Finding 12** -- the opponent does not rate the board as your ratings
  mirrored. Two teams' views of the same 25 matchups correlated at r = -0.049.
  So we cannot predict *which* assignment they will steer toward.
* **Finding 13** -- they do, however, optimise. The opposing team achieved the
  single best assignment available to them out of all 120.

Together those say: model the opponent as an optimiser whose objective you
cannot see. The defensible response is not prediction but a **guarantee**. The
floor computed here is a hard lower bound on your round total over every
completion of the board. It holds regardless of the pairing protocol, and
regardless of what the opponent believes, because it minimises over *all*
remaining assignments -- a superset of those any protocol can actually reach.

That makes the floor conservative by construction: the true worst case under a
specific protocol is never lower, and may be higher. It is a floor you are
guaranteed to clear, not a prediction of where you will land.

Terminology
-----------
floor
    Worst round total over every completion. Guaranteed.
ceiling
    Best round total over every completion. Requires the opponent to cooperate.
spread
    ``ceiling - floor``. How much the remaining pairing decisions are worth at
    all. A spread of zero means the round is already determined.
tau
    The dead-even line, ``games * (rating_min + rating_max) / 2``. Derived in
    ``docs/SCORING_MATHEMATICS.md`` section 1.4. A round is won strictly above it.

Both extremes are exact solutions of a linear assignment problem, solved with
the Kuhn-Munkres/Jonker-Volgenant O(n^3) method. No third-party dependency.
"""

from __future__ import annotations

from collections.abc import Sequence
from dataclasses import dataclass

__all__ = [
    "Outlook",
    "CellChoice",
    "DecisionReport",
    "even_threshold",
    "assignment_extremes",
    "board_outlook",
    "cell_outlooks",
    "decision_report",
]

Matrix = Sequence[Sequence[float]]

UNWINNABLE = "unwinnable"
SECURED = "secured"
LIVE = "live"


def even_threshold(games: int, rating_min: float = 1.0, rating_max: float = 5.0) -> float:
    """The dead-even line for a round of ``games`` games on the given scale.

    Scale-independent: a 1-5 board and a 0-100 board both resolve correctly,
    which is what lets the app offer several rating systems over one engine.
    """
    if games < 0:
        raise ValueError("games must be non-negative")
    return games * (rating_min + rating_max) / 2.0


def _hungarian_min(cost: Matrix) -> tuple[float, list[int]]:
    """Minimum-cost perfect assignment. Returns (total, col_for_row).

    Standard O(n^3) shortest-augmenting-path formulation with potentials.
    Rows must not outnumber columns.
    """
    n = len(cost)
    if n == 0:
        return 0.0, []
    m = len(cost[0])
    if n > m:
        raise ValueError("assignment needs at least as many columns as rows")

    inf = float("inf")
    u = [0.0] * (n + 1)
    v = [0.0] * (m + 1)
    p = [0] * (m + 1)
    way = [0] * (m + 1)

    for i in range(1, n + 1):
        p[0] = i
        j0 = 0
        minv = [inf] * (m + 1)
        used = [False] * (m + 1)
        while True:
            used[j0] = True
            i0 = p[j0]
            delta = inf
            j1 = 0
            for j in range(1, m + 1):
                if used[j]:
                    continue
                cur = cost[i0 - 1][j - 1] - u[i0] - v[j]
                if cur < minv[j]:
                    minv[j] = cur
                    way[j] = j0
                if minv[j] < delta:
                    delta = minv[j]
                    j1 = j
            for j in range(m + 1):
                if used[j]:
                    u[p[j]] += delta
                    v[j] -= delta
                else:
                    minv[j] -= delta
            j0 = j1
            if p[j0] == 0:
                break
        while j0:
            j1 = way[j0]
            p[j0] = p[j1]
            j0 = j1

    col_for_row = [-1] * n
    for j in range(1, m + 1):
        if p[j]:
            col_for_row[p[j] - 1] = j - 1
    total = sum(cost[i][col_for_row[i]] for i in range(n))
    return total, col_for_row


def assignment_extremes(matrix: Matrix) -> tuple[float, float]:
    """Exact (floor, ceiling) over every perfect assignment of ``matrix``."""
    if not matrix:
        return 0.0, 0.0
    lo, _ = _hungarian_min(matrix)
    hi, _ = _hungarian_min([[-x for x in row] for row in matrix])
    return lo, -hi


@dataclass(frozen=True)
class Outlook:
    """What is guaranteed, what is reachable, and whether the round is live."""

    floor: float
    ceiling: float
    tau: float

    @property
    def spread(self) -> float:
        """How much the remaining pairing decisions can move the result."""
        return self.ceiling - self.floor

    @property
    def floor_margin(self) -> float:
        """Guaranteed margin over the dead-even line. Negative means at risk."""
        return self.floor - self.tau

    @property
    def ceiling_margin(self) -> float:
        """Best reachable margin. Negative means the round cannot be won."""
        return self.ceiling - self.tau

    @property
    def verdict(self) -> str:
        if self.ceiling <= self.tau:
            return UNWINNABLE
        if self.floor > self.tau:
            return SECURED
        return LIVE

    @property
    def is_decided(self) -> bool:
        """True when no remaining pairing decision can change the outcome."""
        return self.verdict != LIVE


def board_outlook(matrix: Matrix, tau: float, committed: float = 0.0) -> Outlook:
    """Outlook for a board, optionally with ``committed`` points already banked.

    ``matrix[i][j]`` is our rating for our player *i* against their player *j*,
    higher being better for us. ``committed`` carries the value of pairings that
    are already locked in, so partial boards work without reshaping the caller's
    data.
    """
    lo, hi = assignment_extremes(matrix)
    return Outlook(floor=lo + committed, ceiling=hi + committed, tau=tau)


def _submatrix(matrix: Matrix, drop_row: int, drop_col: int) -> list[list[float]]:
    return [
        [value for j, value in enumerate(row) if j != drop_col]
        for i, row in enumerate(matrix)
        if i != drop_row
    ]


def cell_outlooks(
    matrix: Matrix, tau: float, committed: float = 0.0
) -> dict[tuple[int, int], Outlook]:
    """Outlook for every individual pairing, assuming that pairing is taken.

    The key is ``(our_index, their_index)``. Each value answers: *if this
    matchup happens, what is then guaranteed and what is still reachable?*
    """
    result: dict[tuple[int, int], Outlook] = {}
    for i, row in enumerate(matrix):
        for j, value in enumerate(row):
            rest = _submatrix(matrix, i, j)
            result[(i, j)] = board_outlook(rest, tau, committed + value)
    return result


@dataclass(frozen=True)
class CellChoice:
    """One candidate pairing, with the outlook that follows from taking it."""

    ours: int
    theirs: int
    value: float
    outlook: Outlook

    def dominates(self, other: CellChoice) -> bool:
        """At least as good on both bounds, and strictly better on one."""
        at_least = (
            self.outlook.floor >= other.outlook.floor
            and self.outlook.ceiling >= other.outlook.ceiling
        )
        strictly = (
            self.outlook.floor > other.outlook.floor or self.outlook.ceiling > other.outlook.ceiling
        )
        return at_least and strictly


@dataclass(frozen=True)
class DecisionReport:
    """The safe choice, the opportunistic choice, and what separates them.

    Measured across 31 real event boards, 89% of candidate pairings are
    *strictly dominated* -- another pairing is at least as good on both bounds
    and better on one. A metric that reports only the ceiling cannot see this,
    and rates a dominated choice identically to the one that dominates it.
    """

    board: Outlook
    safest: CellChoice
    boldest: CellChoice
    frontier: tuple[CellChoice, ...]
    hidden_floor_cost: float

    @property
    def choice_matters(self) -> bool:
        """True when protecting the floor genuinely costs ceiling.

        False means a single pairing dominates: it is the safest *and* the most
        opportunistic, so taking it gives up nothing. That is the common case
        (30 of 31 real boards), and it is precisely the case a ceiling-only
        metric cannot distinguish from giving away several points of floor.
        """
        return len({(c.outlook.floor, c.outlook.ceiling) for c in self.frontier}) > 1

    @property
    def floor_at_stake(self) -> float:
        """Guaranteed points given up by taking the bold choice over the safe one."""
        return self.safest.outlook.floor - self.boldest.outlook.floor

    @property
    def ceiling_at_stake(self) -> float:
        """Upside given up by taking the safe choice over the bold one."""
        return self.boldest.outlook.ceiling - self.safest.outlook.ceiling


def decision_report(matrix: Matrix, tau: float, committed: float = 0.0) -> DecisionReport:
    """Rank pairings by guaranteed floor and by reachable ceiling.

    ``hidden_floor_cost`` is the headline number: among the pairings that a
    ceiling-only metric rates as equally good, it is the spread in guaranteed
    floor between the best and the worst of them. On real boards this averages
    2.39 points and is never zero.

    Ties are broken deterministically -- by the opposite bound, then by index --
    so the recommendation never depends on the order candidates are presented in.
    """
    if not matrix or not matrix[0]:
        raise ValueError("decision_report needs a non-empty board")

    cells = [
        CellChoice(ours=i, theirs=j, value=matrix[i][j], outlook=o)
        for (i, j), o in cell_outlooks(matrix, tau, committed).items()
    ]
    safest = max(cells, key=lambda c: (c.outlook.floor, c.outlook.ceiling, -c.ours, -c.theirs))
    boldest = max(cells, key=lambda c: (c.outlook.ceiling, c.outlook.floor, -c.ours, -c.theirs))

    frontier = tuple(
        sorted(
            (c for c in cells if not any(o.dominates(c) for o in cells)),
            key=lambda c: (-c.outlook.floor, -c.outlook.ceiling, c.ours, c.theirs),
        )
    )

    best_ceiling = max(c.outlook.ceiling for c in cells)
    tied_on_ceiling = [c for c in cells if c.outlook.ceiling == best_ceiling]
    hidden = max(c.outlook.floor for c in tied_on_ceiling) - min(
        c.outlook.floor for c in tied_on_ceiling
    )

    return DecisionReport(
        board=board_outlook(matrix, tau, committed),
        safest=safest,
        boldest=boldest,
        frontier=frontier,
        hidden_floor_cost=hidden,
    )

File: webapp/src/engine/boardAnalysis.ts
/**
 * Board analysis: what is guaranteed, what is reachable, and what a
 * single-number sort cannot see.
 *
 * This is a faithful port of `qtr_pairing_process/board_analysis.py`. The two
 * implementations are pinned to the same test vectors (see boardAnalysis.test.ts
 * and test_board_analysis.py) so the phone and the desktop can never disagree
 * about a number shown to a player mid-event.
 *
 * Convention throughout: `matrix[i][j]` is OUR rating for OUR player `i`
 * against THEIR player `j`, higher being better for us.
 */

import { committedChanceExtremes, probabilityMatrix } from "./winProbability";

export type Matrix = readonly (readonly number[])[];

export const UNWINNABLE = "unwinnable";
export const SECURED = "secured";
export const LIVE = "live";
export type Verdict = typeof UNWINNABLE | typeof SECURED | typeof LIVE;

/**
 * The dead-even line for a round of `games` games on the given scale.
 *
 * Scale-independent: a 1-5 board and a 0-100 board both resolve correctly,
 * which is what lets the app keep offering several rating systems over one
 * engine. Players keep the scale they are comfortable with; the maths does not
 * change underneath them.
 */
export function evenThreshold(games: number, ratingMin = 1, ratingMax = 5): number {
  if (games < 0) throw new Error("games must be non-negative");
  return (games * (ratingMin + ratingMax)) / 2;
}

/**
 * Minimum-cost perfect assignment. Returns [total, colForRow].
 *
 * Standard O(n^3) shortest-augmenting-path formulation with potentials.
 * Rows must not outnumber columns.
 */
export function hungarianMin(cost: Matrix): [number, number[]] {
  const n = cost.length;
  if (n === 0) return [0, []];
  const m = cost[0].length;
  if (n > m) throw new Error("assignment needs at least as many columns as rows");

  const INF = Infinity;
  const u = new Array<number>(n + 1).fill(0);
  const v = new Array<number>(m + 1).fill(0);
  const p = new Array<number>(m + 1).fill(0);
  const way = new Array<number>(m + 1).fill(0);

  for (let i = 1; i <= n; i++) {
    p[0] = i;
    let j0 = 0;
    const minv = new Array<number>(m + 1).fill(INF);
    const used = new Array<boolean>(m + 1).fill(false);
    for (;;) {
      used[j0] = true;
      const i0 = p[j0];
      let delta = INF;
      let j1 = 0;
      for (let j = 1; j <= m; j++) {
        if (used[j]) continue;
        const cur = cost[i0 - 1][j - 1] - u[i0] - v[j];
        if (cur < minv[j]) {
          minv[j] = cur;
          way[j] = j0;
        }
        if (minv[j] < delta) {
          delta = minv[j];
          j1 = j;
        }
      }
      for (let j = 0; j <= m; j++) {
        if (used[j]) {
          u[p[j]] += delta;
          v[j] -= delta;
        } else {
          minv[j] -= delta;
        }
      }
      j0 = j1;
      if (p[j0] === 0) break;
    }
    while (j0) {
      const j1 = way[j0];
      p[j0] = p[j1];
      j0 = j1;
    }
  }

  const colForRow = new Array<number>(n).fill(-1);
  for (let j = 1; j <= m; j++) if (p[j]) colForRow[p[j] - 1] = j - 1;
  let total = 0;
  for (let i = 0; i < n; i++) total += cost[i][colForRow[i]];
  return [total, colForRow];
}

/** Exact [floor, ceiling] over every perfect assignment of `matrix`. */
export function assignmentExtremes(matrix: Matrix): [number, number] {
  if (matrix.length === 0) return [0, 0];
  const [lo] = hungarianMin(matrix);
  const [hi] = hungarianMin(matrix.map((row) => row.map((x) => -x)));
  return [lo, -hi];
}

export class Outlook {
  readonly floor: number;
  readonly ceiling: number;
  readonly tau: number;

  constructor(floor: number, ceiling: number, tau: number) {
    this.floor = floor;
    this.ceiling = ceiling;
    this.tau = tau;
  }

  /** How much the remaining pairing decisions can still move the result. */
  get spread(): number {
    return this.ceiling - this.floor;
  }

  /** Guaranteed margin over the dead-even line. Negative means at risk. */
  get floorMargin(): number {
    return this.floor - this.tau;
  }

  /** Best reachable margin. Negative means the round cannot be won. */
  get ceilingMargin(): number {
    return this.ceiling - this.tau;
  }

  get verdict(): Verdict {
    if (this.ceiling <= this.tau) return UNWINNABLE;
    if (this.floor > this.tau) return SECURED;
    return LIVE;
  }

  /** True when no remaining pairing decision can change the outcome. */
  get isDecided(): boolean {
    return this.verdict !== LIVE;
  }
}

/**
 * Outlook for a board, optionally with `committed` points already banked.
 *
 * The floor minimises over EVERY perfect assignment, which is a superset of
 * what the pairing protocol can actually reach. It is therefore conservative
 * by construction: a guarantee, not a prediction.
 */
export function boardOutlook(matrix: Matrix, tau: number, committed = 0): Outlook {
  const [lo, hi] = assignmentExtremes(matrix);
  return new Outlook(lo + committed, hi + committed, tau);
}

function submatrix(matrix: Matrix, dropRow: number, dropCol: number): number[][] {
  const out: number[][] = [];
  for (let i = 0; i < matrix.length; i++) {
    if (i === dropRow) continue;
    const row: number[] = [];
    for (let j = 0; j < matrix[i].length; j++) {
      if (j === dropCol) continue;
      row.push(matrix[i][j]);
    }
    out.push(row);
  }
  return out;
}

/**
 * Outlook for every individual pairing, assuming that pairing is taken.
 *
 * Keyed `"i,j"`. Answers, for each cell: *if this matchup happens, what is
 * then guaranteed and what is still reachable?*
 */
export function cellOutlooks(matrix: Matrix, tau: number, committed = 0): Map<string, Outlook> {
  const result = new Map<string, Outlook>();
  for (let i = 0; i < matrix.length; i++) {
    for (let j = 0; j < matrix[i].length; j++) {
      const rest = submatrix(matrix, i, j);
      const banked = committed + matrix[i][j];
      result.set(
        `${i},${j}`,
        rest.length === 0 ? new Outlook(banked, banked, tau) : boardOutlook(rest, tau, banked),
      );
    }
  }
  return result;
}

export interface CellChoice {
  ours: number;
  theirs: number;
  value: number;
  outlook: Outlook;
}

/** At least as good on both bounds, and strictly better on one. */
export function dominates(a: CellChoice, b: CellChoice): boolean {
  const atLeast = a.outlook.floor >= b.outlook.floor && a.outlook.ceiling >= b.outlook.ceiling;
  const strictly = a.outlook.floor > b.outlook.floor || a.outlook.ceiling > b.outlook.ceiling;
  return atLeast && strictly;
}

export interface DecisionReport {
  board: Outlook;
  safest: CellChoice;
  boldest: CellChoice;
  frontier: CellChoice[];
  /**
   * Among the pairings a ceiling-only metric rates as equally good, the spread
   * in guaranteed floor between the best and the worst of them. Measured across
   * 31 real event boards this averages 2.39 points, reaches 4.0, and is never
   * zero.
   */
  hiddenFloorCost: number;
  /** True when protecting the floor genuinely costs ceiling. */
  choiceMatters: boolean;
  /** Guaranteed points given up by taking the bold choice over the safe one. */
  floorAtStake: number;
  /** Upside given up by taking the safe choice over the bold one. */
  ceilingAtStake: number;
}

/**
 * Rank pairings by guaranteed floor and by reachable ceiling.
 *
 * Ties are broken deterministically -- by the opposite bound, then by index --
 * so a recommendation never depends on the order candidates are presented in.
 */
export function decisionReport(matrix: Matrix, tau: number, committed = 0): DecisionReport {
  if (matrix.length === 0 || matrix[0].length === 0) {
    throw new Error("decisionReport needs a non-empty board");
  }

  const cells: CellChoice[] = [];
  for (const [key, outlook] of cellOutlooks(matrix, tau, committed)) {
    const [i, j] = key.split(",").map(Number);
    cells.push({ ours: i, theirs: j, value: matrix[i][j], outlook });
  }

  const pick = (primary: (c: CellChoice) => number, secondary: (c: CellChoice) => number) =>
    cells.reduce((best, c) => {
      const a = [primary(c), secondary(c), -c.ours, -c.theirs];
      const b = [primary(best), secondary(best), -best.ours, -best.theirs];
      for (let k = 0; k < a.length; k++) {
        if (a[k] !== b[k]) return a[k] > b[k] ? c : best;
      }
      return best;
    });

  const safest = pick(
    (c) => c.outlook.floor,
    (c) => c.outlook.ceiling,
  );
  const boldest = pick(
    (c) => c.outlook.ceiling,
    (c) => c.outlook.floor,
  );

  const frontier = cells
    .filter((c) => !cells.some((o) => dominates(o, c)))
    .sort(
      (a, b) =>
        b.outlook.floor - a.outlook.floor ||
        b.outlook.ceiling - a.outlook.ceiling ||
        a.ours - b.ours ||
        a.theirs - b.theirs,
    );

  const bestCeiling = Math.max(...cells.map((c) => c.outlook.ceiling));
  const tied = cells.filter((c) => c.outlook.ceiling === bestCeiling);
  const hiddenFloorCost =
    Math.max(...tied.map((c) => c.outlook.floor)) - Math.min(...tied.map((c) => c.outlook.floor));

  const distinct = new Set(frontier.map((c) => `${c.outlook.floor}|${c.outlook.ceiling}`));

  return {
    board: boardOutlook(matrix, tau, committed),
    safest,
    boldest,
    frontier,
    hiddenFloorCost,
    choiceMatters: distinct.size > 1,
    floorAtStake: safest.outlook.floor - boldest.outlook.floor,
    ceilingAtStake: boldest.outlook.ceiling - safest.outlook.ceiling,
  };
}

export interface CellChance {
  ours: number;
  theirs: number;
  floor: number;
  ceiling: number;
}

export interface DecisionReportChance {
  safest: CellChance;
  boldest: CellChance;
  /** Guaranteed round-win chance given up by taking the bold cell over the safe one. */
  floorAtStake: number;
  /** Upside chance given up by taking the safe cell over the bold one. */
  ceilingAtStake: number;
  /**
   * The chance-valued twin of `DecisionReport.hiddenFloorCost`: the spread in
   * guaranteed round-win chance across the very pairings a points ceiling rates
   * as equal.
   */
  hiddenFloorCost: number;
}

/**
 * `decisionReport` re-expressed in round-win chance rather than points.
 *
 * The RECOMMENDATION stays points-driven: which cell is "safest" and which is
 * "boldest", and which pairings a ceiling-only metric calls tied, are decided by
 * the same `decisionReport` the panel already trusts. Only the CONSEQUENCES --
 * the floors, ceilings, and the gaps between them -- are recomputed in chance,
 * because a captain deciding whether the trade-off is worth it wants it in the
 * currency that takes the round, not in points.
 *
 * That split is deliberate: `committedChanceExtremes` bounds are not a sum, so
 * the cell that is safest in chance need not be the cell that is safest in
 * points. Re-picking here would quietly recommend a different pairing than the
 * points cards name, and the two currencies are meant to describe ONE decision.
 */
export function decisionReportChance(
  matrix: Matrix,
  tau: number,
  ratingMin = 1,
  ratingMax = 5,
): DecisionReportChance {
  const report = decisionReport(matrix, tau);
  const probs = probabilityMatrix(matrix, ratingMin, ratingMax);

  const chanceAt = (ours: number, theirs: number): CellChance => {
    const [floor, ceiling] = committedChanceExtremes(probs, ours, theirs);
    return { ours, theirs, floor, ceiling };
  };

  const safest = chanceAt(report.safest.ours, report.safest.theirs);
  const boldest = chanceAt(report.boldest.ours, report.boldest.theirs);

  // The same tied set decisionReport measured: cells whose POINTS ceiling ties
  // the board maximum. Recomputed here because decisionReport does not expose
  // it, at the cost of one more cellOutlooks pass (ten sorts of five numbers).
  const outlooks = cellOutlooks(matrix, tau);
  const cells = [...outlooks].map(([key, o]) => {
    const [i, j] = key.split(",").map(Number);
    return { ours: i, theirs: j, ceiling: o.ceiling };
  });
  const bestCeiling = Math.max(...cells.map((c) => c.ceiling));
  const tiedFloors = cells
    .filter((c) => c.ceiling === bestCeiling)
    .map((c) => committedChanceExtremes(probs, c.ours, c.theirs)[0]);

  return {
    safest,
    boldest,
    floorAtStake: safest.floor - boldest.floor,
    ceilingAtStake: boldest.ceiling - safest.ceiling,
    hiddenFloorCost: Math.max(...tiedFloors) - Math.min(...tiedFloors),
  };
}
